File: office_server/src/main/kotlin/foatto/office/cReminder.java
//package foatto.office;
//
//import foatto.core_server.app.server.cStandart;
//import foatto.core_server.app.server.column.iColumn;
//import foatto.core_server.app.server.data.DataBoolean;
//import foatto.core_server.app.server.data.DataDate;
//import foatto.core_server.app.server.data.DataTime;
//import foatto.core_server.app.server.data.iData;
//import foatto.core_client.app.table.CoreTableCellInfo;
//
//import java.util.GregorianCalendar;
//import java.util.HashMap;
//import java.util.HashSet;
//
//public class cReminder extends cStandart {
//
//    private GregorianCalendar toDay = new GregorianCalendar();
//
//    protected StringBuilder addSQLWhere( HashSet<String> hsTableRenameList ) {
//        StringBuilder sb = super.addSQLWhere( hsTableRenameList );
//
//        mReminder m = (mReminder) model;
//        if( m.getType() != -1 )
//            sb.append( " AND " ).append( renameTableName( hsTableRenameList, m.getColumnType().getTableName() ) ).append( '.' )
//              .append( m.getColumnType().getFieldName( 0 ) ).append( " = " ).append( m.getType() );
//
//        return sb;
//    }
//
//    protected CoreTableCellInfo setTableGroupColumnStyle( HashMap<iColumn,iData> hmColumnData, iColumn column, CoreTableCellInfo tci ) {
//        super.setTableGroupColumnStyle( hmColumnData, column, tci );
//
//        mReminder mr = (mReminder) model;
//        if( column.equals( mr.getColumnDate() ) ) {
//            GregorianCalendar gc = ( (DataDate) hmColumnData.get( mr.getColumnDate() ) ).getValue();
//            gc.add( GregorianCalendar.DAY_OF_MONTH, 1 );
//            if( gc.before( toDay ) ) {
//                tci.foreColorType = CoreTableCellInfo.FORE_COLOR_TYPE_DEFINED;
//                tci.foreColor = TABLE_CELL_FORE_COLOR_CRITICAL;
//            }
//        }
//        else if( column.equals( mr.getColumnTime() ) ) {
//            DataDate date = (DataDate) hmColumnData.get( mr.getColumnDate() );
//            DataTime time = (DataTime) hmColumnData.get( mr.getColumnTime() );
//            GregorianCalendar gc = new GregorianCalendar( date.getYear(), date.getMonth() - 1, date.getDay(),
//                                                          time.getHour(), time.getMinute(), 0 );
//            if( gc.before( toDay ) ) {
//                tci.foreColorType = CoreTableCellInfo.FORE_COLOR_TYPE_DEFINED;
//                tci.foreColor = TABLE_CELL_FORE_COLOR_CRITICAL;
//            }
//        }
//        return tci;
//    }
//
//    protected void getTableColumnStyle( int rowNo, boolean isNewRow, HashMap<iColumn,iData> hmColumnData, iColumn column, CoreTableCellInfo tci ) {
//        super.getTableColumnStyle( rowNo, isNewRow, hmColumnData, column, tci );
//
//        mReminder mr = (mReminder) model;
//        if( ( (DataBoolean) hmColumnData.get( mr.getColumnArchive() ) ).getValue() ) {
//            tci.foreColorType = CoreTableCellInfo.FORE_COLOR_TYPE_DEFINED;
//            tci.foreColor = TABLE_CELL_FORE_COLOR_DISABLED;
//        }
//    }
//}

File: office_server/src/main/kotlin/foatto/office/cReminder.kt
package foatto.office

import foatto.core.link.TableCell
import foatto.core.link.TableCellForeColorType
import foatto.core_server.app.server.cStandart
import foatto.core_server.app.server.column.iColumn
import foatto.core_server.app.server.data.DataDate3Int
import foatto.core_server.app.server.data.DataTime3Int
import foatto.core_server.app.server.data.iData
import java.time.LocalDate
import java.time.LocalDateTime

class cReminder : cStandart() {

    override fun addSQLWhere(hsTableRenameList: Set<String>): String {
        val tn = renameTableName(hsTableRenameList, model.tableName)
        val m = model as mReminder

        return super.addSQLWhere(hsTableRenameList) +
            if (m.type != -1) {
                """
                    AND $tn.${m.columnType.getFieldName()} = ${m.type}
                """
            } else {
                ""
            }
    }

    override fun setTableGroupColumnStyle(hmColumnData: Map<iColumn, iData>, column: iColumn, tci: TableCell): TableCell {
        super.setTableGroupColumnStyle(hmColumnData, column, tci)

        val mr = model as mReminder
        if (column == mr.columnDate) {
            val localDate = (hmColumnData[mr.columnDate] as DataDate3Int).localDate
            localDate.plusDays(1)
            if (localDate.isBefore(LocalDate.now())) {
                tci.foreColorType = TableCellForeColorType.DEFINED
                tci.foreColor = TABLE_CELL_FORE_COLOR_CRITICAL
            }
        } else if (column == mr.columnTime) {
            val localDate = (hmColumnData[mr.columnDate] as DataDate3Int).localDate
            val localTime = (hmColumnData[mr.columnTime] as DataTime3Int).localTime
            val dateTime = LocalDateTime.of(localDate, localTime)
            if (dateTime.isBefore(LocalDateTime.now())) {
                tci.foreColorType = TableCellForeColorType.DEFINED
                tci.foreColor = TABLE_CELL_FORE_COLOR_CRITICAL
            }
        }
        return tci
    }

}
